File: src/Soel/Pipeline.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE RecordWildCards   #-}

module Soel.Pipeline
  ( AmbiguityMode(..)
  , CompileOptions(..)
  , RunOptions(..)
  , CheckOptions(..)
  , runCompile
  , runRun
  , runCheck
  , runRepair
  ) where

import Control.Monad.Except (throwError)
import Control.Monad.IO.Class (liftIO)
import Data.Aeson (encode, object, (.=))
import qualified Data.ByteString.Lazy.Char8 as BL8
import Data.Text (Text)
import qualified Data.Text as T
import qualified Data.Text.IO as TIO
import System.FilePath (replaceExtension)
import System.IO (stderr)

import Soel.Types (App, SoelError(..))
import Soel.IR.Types
import Soel.IR.Transform (transformToCodeIR)
import Soel.Stages.Reader (readSource)
import Soel.Stages.SemanticEncoder (semanticEncode)
import Soel.Stages.AmbiguityDetector (detectAmbiguities, emitDiagnostics)
import Soel.Stages.Dialog (resolveAmbiguities, autoResolveAll)
import Soel.Stages.Codegen (generateHaskell, CodegenResult(..))
import Soel.Stages.Writer (writeHaskell, writeIR, WriteOptions(..))
import Soel.Stages.GHC (ghcCompile, ghcRun, GHCResult(..))
import Soel.Stages.Repair (repairLoop)
import Soel.Cache.Store (getCached, setCache)
import Soel.Utils.Logger

-- | How to handle semantic ambiguities.
data AmbiguityMode = Strict | Dialog | Lenient
  deriving (Show, Eq)

-- | Options for the compile command.
data CompileOptions = CompileOptions
  { coOutput       :: !(Maybe FilePath)
  , coIrOnly       :: !Bool
  , coAmbiguityMode :: !AmbiguityMode
  } deriving (Show)

-- | Options for the run command.
data RunOptions = RunOptions
  { roAmbiguityMode :: !AmbiguityMode
  } deriving (Show)

-- | Options for the check command.
data CheckOptions = CheckOptions
  deriving (Show)

-- ─── Compile ────────────────────────────────────────────────────────

runCompile :: FilePath -> CompileOptions -> App ()
runCompile filePath CompileOptions{..} = do
  -- Stage 1: Read
  liftIO $ logStage "Reading source"
  source <- readSource filePath

  -- Stage 2: Semantic Encode (with cache)
  let narrativeCacheKey = "narrative-" <> sfName source
  mCachedNarrative <- getCached narrativeCacheKey (sfHash source)
  narrativeIR <- case (mCachedNarrative :: Maybe NarrativeIR) of
    Just nir -> do
      liftIO $ logInfo "Using cached narrative IR"
      pure nir
    Nothing -> do
      nir <- semanticEncode (sfContent source)
      setCache narrativeCacheKey (sfHash source) nir
      pure nir

  -- Stage 2b: Transform to Code IR (with cache)
  let codeCacheKey = "code-" <> sfName source
  mCachedCode <- getCached codeCacheKey (sfHash source)
  codeIR <- case (mCachedCode :: Maybe CodeIR) of
    Just cir -> do
      liftIO $ logInfo "Using cached code IR"
      pure cir
    Nothing -> do
      cir <- transformToCodeIR narrativeIR (sfContent source)
      setCache codeCacheKey (sfHash source) cir
      pure cir

  -- If --ir-only, output the IR and stop
  if coIrOnly
    then do
      let irVal = object
                    [ "narrative" .= narrativeIR
                    , "code"      .= codeIR
                    ]
      _ <- writeIR irVal (sfPath source)
      liftIO $ BL8.putStrLn (encode irVal)
      pure ()
    else do
      -- Stage 3: Detect Ambiguities
      ambResult <- detectAmbiguities narrativeIR codeIR

      -- Stage 4: Resolve based on mode
      resolved <- case coAmbiguityMode of
        Strict -> do
          emitDiagnostics ambResult (sfPath source) (sfContent source)
          -- If we get here, only warnings remain
          pure $ autoResolveAll (arWarnings ambResult)

        Dialog -> do
          -- Print warnings first
          liftIO $ mapM_ (TIO.hPutStr stderr . formatDiagnostic . ambiguityDiag (sfPath source) (sfContent source))
            (arWarnings ambResult)
          resolveAmbiguities (arAll ambResult) (sfContent source)

        Lenient -> do
          let allAmbs = arAll ambResult
          if not (null allAmbs)
            then do
              liftIO $ mapM_ (TIO.hPutStr stderr . formatDiagnostic . ambiguityDiag (sfPath source) (sfContent source))
                allAmbs
              let e = length (arErrors ambResult)
                  w = length (arWarnings ambResult)
              liftIO $ logWarn $ "Auto-resolving " <> T.pack (show e) <> " error(s) and "
                              <> T.pack (show w) <> " warning(s) in lenient mode"
            else pure ()
          pure $ autoResolveAll allAmbs

      -- Stage 5: Generate Haskell
      CodegenResult{..} <- generateHaskell codeIR resolved (sfContent source)

      -- Stage 6: Write
      hsPath <- writeHaskell crHaskellSource WriteOptions
        { woSourcePath = sfPath source
        , woOutputPath = coOutput
        }

      liftIO $ logSuccess $ "Compilation complete: " <> T.pack hsPath

-- ─── Run ────────────────────────────────────────────────────────────

runRun :: FilePath -> RunOptions -> App ()
runRun filePath RunOptions{..} = do
  let hsPath = replaceExtension filePath ".hs"

  runCompile filePath CompileOptions
    { coOutput        = Just hsPath
    , coIrOnly        = False
    , coAmbiguityMode = roAmbiguityMode
    }

  -- Stage 7: GHC compile + run
  result <- ghcCompile hsPath
  ghcRun (grExecutablePath result)

-- ─── Check ──────────────────────────────────────────────────────────

runCheck :: FilePath -> CheckOptions -> App ()
runCheck filePath _ = do
  liftIO $ logStage "Reading source"
  source <- readSource filePath

  narrativeIR <- semanticEncode (sfContent source)
  codeIR <- transformToCodeIR narrativeIR (sfContent source)
  ambResult <- detectAmbiguities narrativeIR codeIR

  if null (arAll ambResult)
    then liftIO $ logSuccess "No semantic issues detected"
    else do
      liftIO $ mapM_ (TIO.hPutStr stderr . formatDiagnostic . ambiguityDiag (sfPath source) (sfContent source))
        (arAll ambResult)

      let e = length (arErrors ambResult)
          w = length (arWarnings ambResult)
      liftIO $ logError $ T.pack (show e) <> " error(s), " <> T.pack (show w) <> " warning(s)"

      if e > 0
        then throwError $ SemanticAmbiguityError $
               T.pack (show e) <> " semantic error(s), " <> T.pack (show w) <> " warning(s)"
        else pure ()

-- ─── Repair ─────────────────────────────────────────────────────────

runRepair :: FilePath -> App ()
runRepair filePath = repairLoop filePath

-- ─── Helpers ────────────────────────────────────────────────────────

-- | Build diagnostic options from an ambiguity.
ambiguityDiag :: FilePath -> Text -> Ambiguity -> DiagnosticOpts
ambiguityDiag fp sourceText amb = DiagnosticOpts
  { diagFile       = fp
  , diagSeverity   = severityText (ambSeverity amb)
  , diagId         = ambId amb
  , diagCategory   = categoryText (ambCategory amb)
  , diagMessage    = ambDescription amb
  , diagSpan       = ambSourceSpan amb
  , diagSourceText = Just sourceText
  , diagOptions    = map (\o -> (aoLabel o, aoConfidence o)) (ambOptions amb)
  }

File: src/pipeline.ts
import { resolve } from 'node:path';
import type { SoelConfig } from './config.js';
import type { NarrativeIR, CodeIR, Ambiguity } from './ir/types.js';
import { readSource } from './stages/reader.js';
import { semanticEncode } from './stages/semantic-encoder.js';
import { transformToCodeIR } from './ir/transform.js';
import { detectAmbiguities, emitDiagnostics } from './stages/ambiguity-detector.js';
import { resolveAmbiguities, autoResolveAll } from './stages/dialog.js';
import { generateHaskell } from './stages/codegen.js';
import { writeHaskell, writeIR } from './stages/writer.js';
import { ghcCompile, ghcRun } from './stages/ghc.js';
import { getCached, setCache } from './cache/store.js';
import { repairLoop } from './stages/repair.js';
import { log } from './utils/logger.js';
import { formatDiagnostic } from './utils/logger.js';
import chalk from 'chalk';

export type AmbiguityMode = 'strict' | 'dialog' | 'lenient';

// ─── Compile ────────────────────────────────────────────────────────

export interface CompileOptions {
  output?: string;
  irOnly?: boolean;
  ambiguityMode: AmbiguityMode;
  config: SoelConfig;
}

export async function runCompile(
  filePath: string,
  opts: CompileOptions
): Promise<void> {
  // Stage 1: Read
  log.stage('Reading source');
  const source = readSource(filePath);

  // Check cache for narrative IR
  let narrativeIR = getCached<NarrativeIR>(`narrative-${source.name}`, source.hash);

  if (!narrativeIR) {
    // Stage 2: Semantic Encode
    narrativeIR = await semanticEncode(source.content, opts.config);
    setCache(`narrative-${source.name}`, source.hash, narrativeIR);
  } else {
    log.info('Using cached narrative IR');
  }

  // Stage 2b: Transform to Code IR
  let codeIR = getCached<CodeIR>(`code-${source.name}`, source.hash);

  if (!codeIR) {
    codeIR = await transformToCodeIR(narrativeIR, source.content, opts.config);
    setCache(`code-${source.name}`, source.hash, codeIR);
  } else {
    log.info('Using cached code IR');
  }

  // If --ir-only, output the IR and stop
  if (opts.irOnly) {
    writeIR({ narrative: narrativeIR, code: codeIR }, source.path);
    console.log(JSON.stringify({ narrative: narrativeIR, code: codeIR }, null, 2));
    return;
  }

  // Stage 3: Detect Ambiguities
  const ambResult = detectAmbiguities(narrativeIR, codeIR, opts.config);

  // Stage 4: Resolve Ambiguities — behavior depends on mode
  let resolved: Ambiguity[];

  switch (opts.ambiguityMode) {
    case 'strict':
      // Print diagnostics for all, throw on errors — like a traditional compiler
      emitDiagnostics(ambResult, source.path, source.content);
      // If we get here, there were only warnings — auto-resolve them
      resolved = autoResolveAll(ambResult.warnings);
      break;

    case 'dialog':
      // Print warnings, then enter interactive dialog for everything
      for (const w of ambResult.warnings) {
        console.error(formatDiagnostic({
          file: source.path,
          severity: 'warning',
          id: w.id,
          category: w.category,
          message: w.description,
          span: w.sourceSpan,
          sourceText: source.content,
          options: w.options,
        }));
      }
      resolved = await resolveAmbiguities(ambResult.all, source.content, opts.config);
      break;

    case 'lenient':
      // Print warnings to stderr, auto-resolve everything silently
      if (ambResult.warnings.length > 0 || ambResult.errors.length > 0) {
        for (const amb of ambResult.all) {
          console.error(formatDiagnostic({
            file: source.path,
            severity: amb.severity,
            id: amb.id,
            category: amb.category,
            message: amb.description,
            span: amb.sourceSpan,
            sourceText: source.content,
            options: amb.options,
          }));
        }
        const e = ambResult.errors.length;
        const w = ambResult.warnings.length;
        log.warn(`Auto-resolving ${e} error(s) and ${w} warning(s) in lenient mode`);
      }
      resolved = autoResolveAll(ambResult.all);
      break;
  }

  // Stage 5: Generate Haskell
  const { haskellSource } = await generateHaskell(
    codeIR,
    resolved,
    source.content,
    opts.config
  );

  // Stage 6: Write
  const hsPath = writeHaskell(haskellSource, {
    sourcePath: source.path,
    outputPath: opts.output,
  });

  log.success(`Compilation complete: ${hsPath}`);
}

// ─── Run ────────────────────────────────────────────────────────────

export interface RunOptions {
  ambiguityMode: AmbiguityMode;
  config: SoelConfig;
}

export async function runRun(
  filePath: string,
  opts: RunOptions
): Promise<void> {
  const hsPath = resolve(filePath).replace(/\.soel$/, '.hs');

  await runCompile(filePath, {
    ambiguityMode: opts.ambiguityMode,
    config: opts.config,
    output: hsPath,
  });

  // Stage 7: GHC compile + run
  const { executablePath } = await ghcCompile(hsPath, opts.config);
  await ghcRun(executablePath);
}

// ─── Check ──────────────────────────────────────────────────────────

export interface CheckOptions {
  config: SoelConfig;
}

export async function runCheck(
  filePath: string,
  opts: CheckOptions
): Promise<void> {
  log.stage('Reading source');
  const source = readSource(filePath);

  const narrativeIR = await semanticEncode(source.content, opts.config);
  const codeIR = await transformToCodeIR(narrativeIR, source.content, opts.config);
  const ambResult = detectAmbiguities(narrativeIR, codeIR, opts.config);

  if (ambResult.all.length === 0) {
    log.success('No semantic issues detected');
    return;
  }

  for (const amb of ambResult.all) {
    console.error(formatDiagnostic({
      file: source.path,
      severity: amb.severity,
      id: amb.id,
      category: amb.category,
      message: amb.description,
      span: amb.sourceSpan,
      sourceText: source.content,
      options: amb.options,
    }));
  }

  const e = ambResult.errors.length;
  const w = ambResult.warnings.length;
  console.error(
    chalk.bold(`${e} error(s), ${w} warning(s)`)
  );

  if (e > 0) {
    process.exitCode = 1;
  }
}

// ─── Repair ─────────────────────────────────────────────────────────

export interface RepairOptions {
  config: SoelConfig;
}

export async function runRepair(
  filePath: string,
  opts: RepairOptions
): Promise<void> {
  await repairLoop(filePath, opts.config);
}
